src/txn/txn_ckpt.c

```diff
@@ -2821,3 +2828,344 @@ __checkpoint_timing_stress(WT_SESSION_IMPL *session, uint64_t flag, struct times
+    empty = TAILQ_EMPTY(&ckpt_threads->work_qh);
+    __wt_spin_unlock(session, &ckpt_threads->work_lock);
+
+    return (empty);
+}
+
+/*
+ * __checkpoint_reconcile_push_done --
+ *     Push a work done unit to the queue.
+ */
+static int
+__checkpoint_reconcile_push_done(WT_SESSION_IMPL *session, WT_CHECKPOINT_PAGE_TO_RECONCILE *entry)
+{
+    WT_CHECKPOINT_RECONCILE_THREADS *ckpt_threads;
+    uint64_t done_pushed, work_pushed;
+
+    ckpt_threads = S2C(session)->ckpt_reconcile_threads;
+
+    __wt_spin_lock(session, &ckpt_threads->done_lock);
+    TAILQ_INSERT_TAIL(&ckpt_threads->done_qh, entry, q);
+    work_pushed = __wt_atomic_load64(&ckpt_threads->work_pushed);
+    done_pushed = __wt_atomic_add64(&ckpt_threads->done_pushed, 1);
+    __wt_spin_unlock(session, &ckpt_threads->done_lock);
+
+    /* Signal only if it looks like the work is done. */
+    if (__wt_checkpoint_reconcile_queue_empty(session) && work_pushed == done_pushed)
+        __wt_cond_signal(session, ckpt_threads->done_cond);
+
+    return (0);
+}
+
+/*
+ * __wt_checkpoint_reconcile_pop_done --
+ *     Pop a work done unit from the queue. The caller is responsible for freeing it.
+ */
+void
+__wt_checkpoint_reconcile_pop_done(
+  WT_SESSION_IMPL *session, WT_CHECKPOINT_PAGE_TO_RECONCILE **entryp)
+{
+    WT_CHECKPOINT_PAGE_TO_RECONCILE *entry;
+    WT_CHECKPOINT_RECONCILE_THREADS *ckpt_threads;
+    WT_CONNECTION_IMPL *conn;
+
+    *entryp = entry = NULL;
+
+    conn = S2C(session);
+    ckpt_threads = conn->ckpt_reconcile_threads;
+
+    if (TAILQ_EMPTY(&ckpt_threads->done_qh))
+        return;
+
+    __wt_spin_lock(session, &ckpt_threads->done_lock);
+
+    /* Recheck again to confirm whether the queue is empty or not? */
+    if (TAILQ_EMPTY(&ckpt_threads->done_qh)) {
+        __wt_spin_unlock(session, &ckpt_threads->done_lock);
+        return;
+    }
+
+    entry = TAILQ_FIRST(&ckpt_threads->done_qh);
+    TAILQ_REMOVE(&ckpt_threads->done_qh, entry, q);
+    *entryp = entry;
+
+    __wt_spin_unlock(session, &ckpt_threads->done_lock);
+    return;
+}
+
+/*
+ * __checkpoint_reconcile_thread_run --
+ *     Entry function for a checkpoint page reconciliation thread. This is called repeatedly from
+ *     the thread group code so it does not need to loop itself.
+ */
+static int
+__checkpoint_reconcile_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
+{
+    WT_CHECKPOINT_PAGE_TO_RECONCILE *entry;
+    WT_CHECKPOINT_RECONCILE_THREADS *ckpt_threads;
+    WT_DECL_RET;
+    bool signalled;
+
+    WT_UNUSED(thread);
+
+    ckpt_threads = S2C(session)->ckpt_reconcile_threads;
+
+    /* Wait until the next event. */
+    __wt_cond_wait_signal(
+      session, ckpt_threads->work_cond, WT_MILLION, __checkpoint_reconcile_thread_chk, &signalled);
+
+    __checkpoint_reconcile_pop_page(session, &entry);
+    if (entry == NULL)
+        return (0);
+
+    memcpy(&session->txn->snapshot_data, entry->snapshot, sizeof(WT_TXN_SNAPSHOT));
```

Comment:
This is a bit tricksy, since the snapshot is captured in an array that is allocated separately to the WT_TXN_SNAPSHOT structure. This code might need to be careful not to free the WT_TXN_SNAPSHOT::snapshot referred to when the fields are copied across.
I haven't double checked in your change if it's an actual problem, but just capturing a thought as I look through